<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>PLINK cheatsheet</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">PLINK cheatsheet</h1>



<p>The following flags and corresponding tidypopgen functions are based
on plink version 1.9.</p>
<div id="file-management-and-reading-data" class="section level2">
<h2>File management and reading data:</h2>
<table>
<colgroup>
<col width="36%" />
<col width="63%" />
</colgroup>
<thead>
<tr class="header">
<th>PLINK 1.9</th>
<th><code>tidypopgen</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>–make-bed –out</td>
<td><code>gt_as_plink(data, file = my_file, type = &quot;bed&quot;)</code></td>
</tr>
<tr class="even">
<td>–recode</td>
<td><code>gt_as_plink(data, file = my_file, type = &quot;ped&quot;)</code></td>
</tr>
<tr class="odd">
<td>–recode vcf</td>
<td><code>gt_as_vcf(data, file = my_file)</code></td>
</tr>
<tr class="even">
<td>–allele1234 and –alleleACGT</td>
<td>See <code>gen_tibble()</code> parameter ‘valid_alleles’</td>
</tr>
</tbody>
</table>
<p>PLINK flags –update-alleles, –allele1234, and –alleleACGT, all alter
the coding of alleles. In <code>tidypopgen</code>, valid alleles are
supplied when reading in a <code>gen_tibble</code>.</p>
</div>
<div id="quality-control" class="section level2">
<h2>Quality control:</h2>
<table>
<thead>
<tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>–maf</td>
<td><code>data %&gt;% loci_maf()</code></td>
</tr>
<tr class="even">
<td>–geno</td>
<td><code>data %&gt;% loci_missingness()</code></td>
</tr>
<tr class="odd">
<td>–hwe</td>
<td><code>data %&gt;% loci_hwe()</code></td>
</tr>
<tr class="even">
<td>–freq –a2-allele</td>
<td><code>data %&gt;% loci_alt_freq()</code></td>
</tr>
<tr class="odd">
<td>–mind</td>
<td><code>data %&gt;% indiv_missingness()</code></td>
</tr>
<tr class="even">
<td>–het</td>
<td><code>data %&gt;% indiv_het_obs()</code></td>
</tr>
</tbody>
</table>
<p>To filter out variants in <code>tidypopgen</code>, in a similar way
to PLINK flags such as –extract or –autosome, it is necessary to use the
<code>gen_tibble</code> with <code>select_loci_if()</code>. For
example:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>data <span class="sc">%&gt;%</span> <span class="fu">select_loci_if</span>(<span class="fu">loci_chromosomes</span>(genotypes) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>))</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#&gt; # A gen_tibble: 15 loci</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#&gt; # A tibble:     5 × 3</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#&gt;   id    population  genotypes</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;      &lt;vctr_SNP&gt;</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt; 1 GRC24 pop_a       [0,0,...]</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt; 2 GRC25 pop_a       [0,0,...]</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt; 3 GRC26 pop_a       [0,1,...]</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#&gt; 4 GRC27 pop_a       [0,0,...]</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#&gt; 5 GRC28 pop_a       [0,0,...]</span></span></code></pre></div>
<p>will select autosomal loci in the same way as –autosome. Or
alternatively:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>my_snps <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;rs4477212&quot;</span>, <span class="st">&quot;rs3094315&quot;</span>, <span class="st">&quot;rs3131972&quot;</span>, <span class="st">&quot;rs12124819&quot;</span>, <span class="st">&quot;rs11240777&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="fu">select_loci_if</span>(<span class="fu">loci_names</span>(genotypes) <span class="sc">%in%</span> my_snps) <span class="sc">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="fu">show_loci</span>()</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; # A tibble: 5 × 8</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt;   big_index name  chromosome position genetic_dist allele_ref allele_alt chr_int</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt;       &lt;int&gt; &lt;chr&gt; &lt;chr&gt;         &lt;int&gt;        &lt;int&gt; &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; 1         1 rs44… 1             82154            0 A          &lt;NA&gt;             1</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; 2         2 rs30… 1            752566            0 A          G                1</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; 3         3 rs31… 1            752721            0 G          A                1</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; 4         4 rs12… 1            776546            0 A          &lt;NA&gt;             1</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; 5         5 rs11… 1            798959            0 G          A                1</span></span></code></pre></div>
<p>will select loci from a previously defined set in the same way as
–extract.</p>
<p>Similarly, to filter out individuals, as might be performed with
–keep in PLINK, requires using filter:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>my_individuals <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;GRC14300079&quot;</span>, <span class="st">&quot;GRC14300142&quot;</span>, <span class="st">&quot;GRC14300159&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(id <span class="sc">%in%</span> my_individuals)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt; # A gen_tibble: 16 loci</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt; # A tibble:     0 × 3</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; # ℹ 3 variables: id &lt;chr&gt;, population &lt;chr&gt;, genotypes &lt;vctr_SNP&gt;</span></span></code></pre></div>
</div>
<div id="handling-linkage" class="section level2">
<h2>Handling linkage</h2>
<p>Linkage disequilibrium is managed through clumping in
<code>tidypopgen</code> with <code>loci_ld_clump()</code>.</p>
<p>This option is similar to the –indep-pairwise flag in PLINK, but
results in a more even distribution of loci when compared to LD
pruning.</p>
<p>To explore why clumping is preferable to pruning, see <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="uri">https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</a></p>
</div>
<div id="quality-control-for-relatedness-king" class="section level2">
<h2>Quality control for relatedness (KING)</h2>
<table>
<thead>
<tr class="header">
<th>KING</th>
<th><code>tidypopgen</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>–kinship</td>
<td><code>pairwise_king()</code></td>
</tr>
<tr class="even">
<td>–distance</td>
<td><code>pairwise_ibs()</code></td>
</tr>
<tr class="odd">
<td>–unrelated</td>
<td><code>filter_high_relatedness()</code></td>
</tr>
</tbody>
</table>
<p><code>pairwise_king()</code> implements the KING-robust estimator of
kinship, equivalent to –kinship in KING. To remove related individuals,
the user can pass the kinship matrix of <code>pairwise_king()</code> and
a relatedness threshold (a numeric KING kinship coefficient) to
<code>filter_high_relatedness()</code>, which will return the largest
possible set of individuals with no relationships above the
threshold.</p>
<p><code>pairwise_king()</code> also forms part of
<code>qc_report_indiv()</code> through the parameter
<code>kings_threshold</code>. To remove related individuals in
<code>qc_report_indiv()</code>, the user can pass either a relatedness
threshold, or a string of either “first” or “second” to remove any first
degree or second degree relationships from the dataset. This second
option is similar to using –unrelated –degree 1 or –unrelated –degree 2
in KING. <code>qc_report_indiv()</code> returns a dataframe including
columns ‘id’ and ‘to_keep’, showing the ID of individuals and a logical
column of whether the individual should be removed to retain the largest
possible set of individuals with no relationships above the
threshold.</p>
</div>
<div id="merging-datasets" class="section level2">
<h2>Merging datasets:</h2>
<table>
<thead>
<tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>–bmerge</td>
<td><code>rbind()</code></td>
</tr>
<tr class="even">
<td>–flip-scan</td>
<td><code>rbind_dry_run()</code></td>
</tr>
<tr class="odd">
<td>–flip</td>
<td>use ‘flip_strand = TRUE’ in <code>rbind()</code></td>
</tr>
</tbody>
</table>
<p>In PLINK, data merging can fail due to strand inconsistencies that
are not addressed prior to merging. PLINK documentation suggests to
users to try a ‘trial flip’ of data to address this, and then to
‘unflip’ any errors that remain. In <code>tidypopgen</code>, when data
are merged with rbind, strand inconsistencies are identified and
automatically flipped, avoiding multiple rounds of flipping before
merging.</p>
<p>PLINK does allow users to identify inconsistencies prior to merging
with –flip-scan, and this functionality is included in the
<code>tidypopgen</code> rbind_dry_run(). rbind_dry_run() reports the
numeric overlap of datasets, alongside the number of SNPs to ‘flip’ in
the new target dataset, as well as the number of ambiguous SNPs.</p>
<p>Data are only merged one set at a time, there is no equivalent to
–merge-list.</p>
</div>
<div id="analysis" class="section level2">
<h2>Analysis:</h2>
<table>
<thead>
<tr class="header">
<th>PLINK</th>
<th><code>tidypopgen</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>–pca</td>
<td>See <code>gt_pca()</code> for pca options</td>
</tr>
<tr class="even">
<td>–fst</td>
<td><code>pairwise_pop_fst()</code> with <code>group_by()</code></td>
</tr>
<tr class="odd">
<td>–homozyg</td>
<td><code>windows_indiv_roh()</code></td>
</tr>
</tbody>
</table>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
